Variable Sensitivity with constant_abstract_objectt

thk123 opened the pull request:
Issue: diffblue/verification-engine-utils#425

Depends on: diffblue/cbmc#472

A WIP on the variable sensitivity framework. It needs some substantial cleanup and obviously some tests (though I suppose I can mostly take the constant propagation ones until we have @owen-jones-diffblue tests. 

Real diff: https://github.com/danpoe/cbmc/pull/1/files/f41a478bca868548913458cddec3ab8a45bf09a2..9a50171a688a1be6fc791e7d08de5fd6f6253ea8

Commit:
Add XML and JSON output to the base of both ai domains and the ai analysis.

Commit:
Make evaluation and/or simplification of conditions/expressions an ai domain operation.

Commit:
Remove the special case handling of --show-intervals.

The option string --show --intervals is more flexible, --show-intervals is now an alias.

Commit:
Constant propagator and intervals regression tests

Where the current precision is insufficient, the tests are marked FUTURE.

Commit:
Implement domain_simplify for the constant domain.

Commit:
Add output_json to the dependence_graph abstract domain.

Commit:
Add utility function to tell if a program is threaded or not.

Commit:
Implement domain_simplify for intervals.

Commit:
Refactor goto-analyzer so that task, analysis, domain and output can be picked independently.

Commit:
Add utility function for remove_unreachable.

Commit:
Catch exceptions thrown in doit() in goto analyzer.

Commit:
Fixes and improvements to the constant and interval domain.

danpoe committed:
Fix merging issues in goto analyzer

danpoe committed:
Fix whitespace to match coding guidelines

danpoe committed:
Fix constant propagator

Four additional regression tests are not expected to work at present which are
now marked as FUTURE.

danpoe committed:
camel case for json

danpoe committed:
disable two-way propagation for now

danpoe committed:
regression test changes

danpoe committed:
Addressing goto analyzer review comments

danpoe committed:
option to ignore unresolved functions in the constant propagator

danpoe committed:
use ai function output in goto-instrument

Commit:
Initial class checkin for the variable sensistivity framework

Commit:
Exclude me - removing final and override on non-virtual method

Commit:
Initial checkin of Martins code

danpoe committed:
more status output

danpoe committed:
constant propagator fix

Commit:
Added hook for running the variable sensitivity

Involved removing messaget inheritance since we require a zero parameter
constructor.

Commit:
Changes required to get sometihng running

Commit:
Added constant_abstract_value implementation

Commit:
Remove top from the map

Commit:
Use variable in the constant_propogation checks

danpoe committed:
add is_bottom() and is_top() to ai_domain_baset

Commit:
Temp checkin to fix the fixed point being found too early

TODO: at the moment it never stops because the merge operation always
returns true since the merge of two AOs doesn't return a value to say if
it changed.

Commit:
Disabling throw other

Commit:
Terminate correctly

Made the domain correclty find the fixed point. The merge operation
returns whether the merge actually changed any values by implementing a
operator== for each of the abstract objects.

Further, when merging a map, if a key is absent from one map then we
remove it from the other unless the map is bottom.

Commit:
Very verbose and debuggy output

Commit:
Temp checkin of the other types we need to support in the AE

Commit:
Adding handlers for the special types

They are just stubs at the moment, returning top as we need the abstract
objects that represent them.

thk123 commented:
@martin-cs, @tautschnig @danpoe  specific areas I'd like general feedback on: 

The way the merge deals with missing entries in the map: https://github.com/danpoe/cbmc/pull/1/files#diff-7260d207dbd5603305c9641edbcd53b5R349

The way I detect and report merges that don't change anything: https://github.com/danpoe/cbmc/pull/1/files#diff-487d940b8f9e1c792514613569de94d4R135

Whether this dynamic cast of references is actually OK: https://github.com/danpoe/cbmc/pull/1/files#diff-4aec1ad827a35a160dcf1e9fcf7f3415R46

thk123 commented:
WIP PR: danpoe/cbmc#1 (targeted to Dan's goto-analyzer branch for now). 

Still to do for this issue:

- [x] implement the stack for assigning to LHS https://github.com/danpoe/cbmc/pull/1/files#diff-7260d207dbd5603305c9641edbcd53b5R144
- [ ] two element classes for array, struct, pointer (boolean, integer?)
- [ ] better stub for assume 
- [ ] messaget?
- [ ] Tidy up 

At the moment there isn't an easy way to output info through the `messaget` interface, I was thinking of making `ai_domain_baset` inherit from `messaget` and when it gets instantiated through `ait`, it has a constructor that takes the `message_handlert`. Is there a reason not to do this?

Commit:
Updated interface for ai_domain_baset

Commit:
Better stub for assigning to non-trivial symbol

Commit:
Created skeleton classes for the array, pointer, struct

These clasess will represent the basis for complex types with helper
methods to read and write to/from them. They however will still be just
a 2 element abstraction (top or bottom)

Commit:
Adding a 2 element implemention of sturct abstract object

thk123 commented:
@martin-cs @danpoe  See this commit for the 2 element implementation of the struct abstract object, does it all look sensible https://github.com/danpoe/cbmc/pull/1/commits/9a50171a688a1be6fc791e7d08de5fd6f6253ea8 ?

tautschnig commented on regression/Makefile:
```diff
@@ -11,3 +11,11 @@ DIRS = ansi-c \
+			$(MAKE) -C "$$dir" clean; \
+		fi; \
+	done;
+
```
May I ask for a separate pull request to have get this cleanup?

tautschnig commented on regression/goto-analyzer/intervals1/test.desc:
```diff
@@ -1,15 +0,0 @@
-^\[main.assertion.7\] file intervals1.c line 26 function main, assertion i>=11: SUCCESS$
-^\[main.assertion.8\] file intervals1.c line 29 function main, assertion j<100: SUCCESS$
---
-^warning: ignoring
```
Why is this one (and others) being removed?

tautschnig commented:
Some overall comments, because I find it rather challenging to work my way through this one:
1. It would be great to have commit messages that explain what this even is about.
2. I find the "abstract" in various parts of the abstract domain rather redundant, unless this is an abstraction of abstract interpretation. Indeed I don't really know where the entry points are as the interplay of the various classes doesn't follow very naturally from their names.
3. Obviously a rebase would be very helpful to reduce the overall size of this PR.

Given the size of this pull request: a pointer to where the actual domain operations are implementation would be very helpful! Thanks!

thk123 commented:
@tautschnig ~Ah sorry thanks for having a look anyway. The real diff is here: https://github.com/danpoe/cbmc/pull/1/files/f41a478bca868548913458cddec3ab8a45bf09a2..9a50171a688a1be6fc791e7d08de5fd6f6253ea8~ rebased so just relevant commits now. Links updated

I linked to a few key places in my previous comment - but I now realise those links were broken by the rebase, I have updated them.

martin-cs commented:
On Tue, 2017-02-21 at 05:06 -0800, Thomas Kiley wrote:
> @martin-cs, @tautschnig @danpoe  specific areas I'd like general feedback on: 
> 
> The way the merge deals with missing entries in the map: https://github.com/danpoe/cbmc/pull/1/files#diff-7260d207dbd5603305c9641edbcd53b5R311

1. Use the sharing map's "iterate over the common elements"
functionality.  @danpoe can advise.

2. A general design decision (which I probably should have mentioned
before) is that we don't bother storing things whose value is \top.
This is because there are a lot of them and it makes it more flexible in
use if missing value == assume top.  Thus the logic for merge should be:

if (this->is_bottom()) {
  return op;
} else if (op.is_bottom()) {
  return *this;
} else {

  return
  forall (objects o that are in both maps)
     this->map[o].merge(op.map[o])
}



> The way I detect and report merges that don't change anything: https://github.com/danpoe/cbmc/pull/1/files#diff-487d940b8f9e1c792514613569de94d4R135

The abstract object merge should give a boolean return value.  The
sharing_map "iterate over the common elements" should aggregate these so
it should be simple to find this case.

> Whether this dynamic cast of references is actually OK: https://github.com/danpoe/cbmc/pull/1/files#diff-4aec1ad827a35a160dcf1e9fcf7f3415R46

Hmmm... I hadn't really thought about testing equality between abstract
objects.  What is this needed for?

martin-cs commented:
On Tue, 2017-02-21 at 10:05 -0800, Thomas Kiley wrote:
> @martin-cs @danpoe  See this commit for the 2 element implementation of the struct abstract object, does it all look sensible https://github.com/danpoe/cbmc/pull/1/commits/9a50171a688a1be6fc791e7d08de5fd6f6253ea8 ?

Conceptually : yes.
Implementation : a few tweaks needed.

1. If the abstract domain for a struct is bottom then the fields are
also bottom.  You're right that it isn't a particularly meaningful thing
to ask but it should work.

2. I might write the write (!) slightly differently:

if (is_top())
 return this;   // In this case it shouldn't break sharing
else
  clone and set to bottom

In the sketch I had a constructor with a flag for 'is top, is bottom'
and I had been hoping we could make the fields that each layer of the
heirarchy adds private.

tautschnig commented:
Thanks a lot for the pointer. What would really help me a lot is a description of the abstract domain - what is bottom, what is top, and what is in between? And how are these meant to be interpreted? That would be nice to see either in a comment or in the commit message. Thus far, I just see lots of objects of different type being created. Thanks!

thk123 commented:
@martin-cs @danpoe To use the sharing map I need to merge in this branch: [danpoe/cbmc:sharing-map-2](https://github.com/danpoe/cbmc/tree/sharing-map-2)? 

Yes have been removing top from the map: https://github.com/danpoe/cbmc/pull/1/files#diff-7260d207dbd5603305c9641edbcd53b5R398 I don't really understand what your code is showing/where it would go? 

The `operator==` is used to find out whether a `abstract_objectt::merge` actually changed anything: https://github.com/danpoe/cbmc/pull/1/files#diff-487d940b8f9e1c792514613569de94d4R141. I'm open to other ways to achieve this.

Commit:
Modifications to struct_abstract_object

Made so the read returns a bottom element if it is bottom.
Made the write return itself if true to faciliate sharing.

Commit:
Readded broken protected

Can't sue type directly when accessing a parent type so need to use the
getter.

martin-cs commented:
On Wed, 2017-02-22 at 02:06 -0800, Michael Tautschnig wrote:
> Thanks a lot for the pointer. What would really help me a lot is a description of the abstract domain - what is bottom, what is top, and what is in between? And how are these meant to be interpreted? That would be nice to see either in a comment or in the commit message. Thus far, I just see lots of objects of different type being created. Thanks!

Please allow me to write something that approximates design
documentation:

goto-analyze   [in danpoe/goto-analyze plus vision for future]
============

Goto-analyze is the tool front-end for CPROVER's classic style abstract
interpretation analyses.  By "classic style" I mean, domains are C++
objects, transformers are computed using C++ functions and the
fix-points are computed by iteration.  This is in contrast to 2LS's
approach where everything is computed using quantifier-elimination.

The analyses are mostly in analyses/ and although they are used in other
places in the code, the goal is that goto-analyze should eventually
provide an executable front-end for all of them.

There are a lot of different analyses and a lot of ways they can be
used.  Goto-analyze has five, largely independent, sets of options:

1. Task : What you do once you've computed the domains.
2. Abstract interpreter : What kind of abstract interpretation you do.
3. Domain : What domain you use.
4. Sensitivity : How that domain handles things like arrays, pointers,
etc.  (This is not implemented yet but see below).
5. Output : What you do with the results.

Formally speaking, 2, 3 and 4 are an artificial distinction as they are
all really parts of the "what domain" question.  However they correspond
to parts of our code architecture, so ... they should stay.

Ideally, the cross product of options should be supported but ... in
practice there will always be ones that are not meaningful.  Those
should give errors.  In addition there are some analyses which are
currently stand alone as they don't fit directly in to this model.  For
example the unwind bounds analysis, which is both the task, the abstract
interpreter and the output.  Where possible these should be integrated /
support the other options.  In the case of the unwind analysis this
means the domain and it's sensitivity should be configurable.


Task
----

Tasks are implemented by the relevant file in goto-analyze/static_*. At
the moment they are each responsible for building the domain / using the
other options.  As much as possible of this code should be shared.  Some
of these inherit from messaget.  They all probably should.

Show : Prints out the domains for inspection or further use.

Verify : Uses the domain to check all assertions in the code, marking
each one "SUCCESS" (the assertion always holds), "FAILURE if
reachable" (the assertion never holds), "UNKNOWN" (the assertion may or
may not hold).  This is implemented using domain_simplify().

Simplify : Generates a new goto program with branch conditions,
assertions, assumptions and assignments simplified using the information
in the domain (again using domain_simplify()).  This task is intended to
be used as a preprocessing step for other analysis.

Invariants : (Not implemented yet) Use the domains to add assume()
statements to the code giving invariants generated from the domain.
These assumes don't reduce the space of possible executions; they make
existing invariants explicit.  Again, intended as a preprocessing step.


Abstract Interpreter
--------------------

This option is effectively about how we compute the fix-point(s) /
which child class of ai_baset we use.  I.E.  ait<domainT> or
concurrency_aware_ait<domainT>, etc.   For migrating / refactor /
unifying with the pointer analysis code we might want a
location_insensitive_ait<domainT> or something but this is not urgent.
We will need a context_aware_ait<domainT>, quite possibly the one
@danpoe has written elsewhere.


Domain
------

Which child of ai_domain_baset we use to represent the abstract state at
each location / implement the transformers.  I expect most of these will
be non-relational (i.e. an abstract object for each variable) due to the
cost of implementing effective non-relational domains in this style vs.
using 2LS.  The exception might be equalities, which we could implement
here.


Output
------

Text, XML, JSON plus some others for specific domains such as dependence
graphs in DOT format.



Variable Sensitivity Domain   [ thk123's branch ]
===========================

There are different ways of handling arrays, structures, unions and
pointers.  Our existing solution basically ignores them which is
imprecise at best and out-right wrong at worst.  For one project we
needed to do better.  We could have implemented a particular way of
handling them in an existing domain, created a new one with it, etc.
This would work but it means duplicate code and it is is inflexible when
the same person / the next person comes along and says "actually, we
really care about the pointer precision but less so the array so could
you just ...".  Thus the idea was to do this properly:

1. Build a "non-relational domain" and allow the abstractions used for
individual variables to be different.

2. Give the user the option of which abstractions are used for structs,
unions, arrays and pointers.  These are the sensitivity options
discussed above.

3. Have the domain options control which kind of abstractions are used
for the individual values, i.e. constants, intervals, etc.


This is implemented in three parts:

abstract_objectt : The base / interface for abstractions of a single
variable.  The interface for these is effectively create (as top,
bottom, from a constant or copy), transform, merge, convert to constant
if possible.  Child classes add additional bits of interface, for
example array abstractions need a "read element" and a "write element"
method, structures need a "read field" and "write field", etc.  These
objects are intended to be immutable and thus operations tend to produce
pointers.  This is so that we can easily produce copy-on-write maps of
them which we need for scaling and performance.  There are also children
of these for the constant abstraction of one value, the interval
abstraction of one value (to be implemented), etc.

abstract_environment : This contains the map from variable names for
abstract_objectt's (the "non-relational" part of the domain).  The map
itself if copy-on-write for performance and scalability but this is all
wrapped up nicely in @danpoe's sharing_map.  The interface here is
evaluate (exprt -> abstract_objectt*), assign (name, abstract_objectt*
-> bool), assume (exprt -> bool) and merge.  It has a factory to build
abstract_objectt* from types or constants but apart from that, doesn't
know anything about which actual abstract_objectt's are being used.  As
long as program variables that are arrays have an abstract_objectt which
has the array interface, and so on for structs, unions, etc. then the
abstractions used for values can be freely mixed and matched in any way
the user can get the factory to build.

variable_sensitivity_domaint : Implements the ai_domain_baset interface
using an abstract_environment.  The only real code here is the
'transform' method which looks at the instruction type and converts that
into calls to eval, assume, assign and merge.

tautschnig commented:
Thank you very much for this nice write-up - which makes me greedy, of course: could we have that included in the source as doxygen? I might not be the last person to ask those questions...

The only extra bit that would be great to have is an example, i.e., a small program with the analysis applied to it. As said, I'm being greedy.

martin-cs commented:
On Wed, 2017-02-22 at 02:14 -0800, Thomas Kiley wrote:
> @martin-cs @danpoe To use the sharing map I need to merge in this branch: [danpoe/cbmc:sharing-map-2](https://github.com/danpoe/cbmc/tree/sharing-map-2)? 

@danpoe knows the current status of these branches...

> Yes have been removing top from the map: https://github.com/danpoe/cbmc/pull/1/files#diff-7260d207dbd5603305c9641edbcd53b5R398 I don't really understand what your code is showing/where it would go? 

I probably wasn't too clear.  My code was the logic for if the
abstract_environment was top or bottom: in full that is:

bool abstract_environmentt::merge(const abstract_environmentt &op)
{

if (this->is_top()) {
  return false;

} else if (op.is_top()) {
  this->set_to_top();   // Which should set a flag and clear the map
  return true;

} else if (this->bottom()) {
  *this = op;     // Copy over the map and the flags
  return true;

} else if (op.bottom()) {
  // Do nothing
  return false;

} else {

// Here we have to actually merge the maps

}
}

Conceptually the merge of maps is building a new map with the set of
keys equal to the intersection of the keys of the two maps and the
values generated by the abstract_objectt::merge of the corresponding
pair of objects.  (Plus the extra caveat that if the merge returns top,
you should skip that key).

The sharing_map should be able to do the difficult work of finding the
intersection and telling when that has actually resulted in a change, as
long as the abstract_objectt::merge does the right thing.

> The `operator==` is used to find out whether a `abstract_objectt::merge` actually changed anything: https://github.com/danpoe/cbmc/pull/1/files#diff-487d940b8f9e1c792514613569de94d4R135. I'm open to other ways to achieve this. 

This should be possible from the sharing_map.  Otherwise... the
semantics get a bit tricky.

martin-cs commented:
On Wed, 2017-02-22 at 04:33 -0800, Michael Tautschnig wrote:
> Thank you very much for this nice write-up - which makes me greedy, of course: could we have that included in the source as doxygen? I might not be the last person to ask those questions...

Yeah... that or the manual...  will do when I have time.  At the moment
@danpoe and @thk123 own the branches that contain the code ...

> The only extra bit that would be great to have is an example, i.e., a small program with the analysis applied to it. As said, I'm being greedy.

At some point I was thinking of doing something horrible to the
internals to generate animated traces / series of pictures of the AI
engine running.  Partially for teaching.  Shouldn't be too hard but
might need the web folks.

Commit:
Implementation of simplify

We try and evaluate conditions and replace them with wether they are
true or not.

Commit:
Don't use static for the map as means wrong this is captured

Commit:
Added handler for logical == expression

Commit:
Cleaned up the variable sensitivity framework

Talked through with Dan what I had done so far and made some
simplifications and improvments to how it works.

Commit:
Corrected the logic for non-trivial assigns
